# REAL RACING 3 VEHICLES - OUTCOME 1

### **Welcome to my 1st part of data science page for Real Racing 3 Vehicles!**

As always, we must begin with importing the necessary modules first.

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

And, of course, read the data frame that contains all vehicles...

**Keep in mind that the dataset of vehicle roster was related to game version 8.0.**

In [47]:
rr3_df = pd.read_csv('real_racing_3_vehicles.csv')

Let's see what we have in the first 10 rows:

In [48]:
rr3_df.head(10)

Manufacturer            Model    PR  Top_Speed(kph)  Acceleration(s)  \
0         Acura          NSX GT3  65.2           273.0             3.01   
1         Acura           ARX-05  78.4           321.0             2.79   
2         Ariel         Atom 3.5  44.4           249.0             2.70   
3         Ariel          Atom V8  61.1           321.0             2.29   
4  Aston Martin              DB9  28.2           294.0             4.50   
5  Aston Martin         Vanquish  33.6           294.0             4.09   
6  Aston Martin    V12 Vantage S  43.8           329.0             3.70   
7  Aston Martin  Vantage AMR Pro  51.3           302.0             3.59   
8  Aston Martin       Vantage 59  40.4           313.0             4.00   
9  Aston Martin     Vantage N430  40.2           313.0             4.09   

   Braking(m)  Grip(g) Class Type  Series Purchase_Cost  Service_Time(min)  \
0        29.5     1.65     R   MR       2      750 Gold              180.0   
1        25.6     1.76     R   MR       1      850 Gold              180.0   
2        30.4     1.00     S   RR       2     255000 R$              140.0   
3        30.4     1.20     S   RR       2      150 Gold              180.0   
4        32.0     1.00     P   FR       4     230000 R$              130.0   
5        31.3     1.01     P   FR       3      110 Gold              150.0   
6        30.4     1.10     P   FR       3     425000 R$              170.0   
7        28.0     1.29     R   FR       2      500 Gold              180.0   
8        30.4     1.07     P   FR       2      400 Gold              180.0   
9        30.1     1.14     P   FR       2      240 Gold              180.0   

   Service_Cost(R$)  Availability  
0            1990.0          True  
1            1997.0          True  
2            1170.0          True  
3            1402.0          True  
4            1132.0          True  
5            1230.0          True  
6            1355.0          True  
7            1415.0          True  
8            1415.0          True  
9            1415.0          True

Let's get the manufacturer and the model name of the 45th vehicle in the data frame:

In [49]:
rr3_df.loc[44, ['Manufacturer', 'Model']]

Manufacturer    Bugatti
Model            Chiron
Name: 44, dtype: object

Well, this returned a Series. But, I want it to be appeared as a plain text:

In [50]:
print(rr3_df.loc[44, 'Manufacturer'] + ' ' + rr3_df.loc[44, 'Model'])

Bugatti Chiron


OK, what are the specs of the vehicle at 125th index?

In [51]:
rr3_df.loc[125, 'PR':'Grip(g)']

PR                 115.8
Top_Speed(kph)       313
Acceleration(s)        2
Braking(m)            17
Grip(g)             3.03
Name: 125, dtype: object

Hmmm, this probably looks like a race car. Could it be from Formula 1?

Anyway, this time, I'd like to see the cost to buy, total service time and cost for the 200th vehicle:

In [52]:
rr3_df.loc[199, ['Purchase_Cost', 'Service_Time(min)', 'Service_Cost(R$)']]

Purchase_Cost        600 Gold
Service_Time(min)         180
Service_Cost(R$)         1870
Name: 199, dtype: object

According to this result, it should be an  expensive sports car that belongs to a well-known manufacturer.

How about the classes and types for vehicles between 251st and 260th index?

In [53]:
rr3_df.loc[251:260, ['Class', 'Type']]

Class Type
251     R   RR
252     S   RR
253     S   RR
254     P   RR
255     P   RR
256     P   MR
257     P   MR
258     S   MR
259     P   MR
260     R   MR

All right, we already have plenty of time to inspect vehicles individually. Now, let's get more detailed info about this data frame.

First of all, these are the columns in the data frame I'm working with:

In [55]:
rr3_df.columns

Index(['Manufacturer', 'Model', 'PR', 'Top_Speed(kph)', 'Acceleration(s)',
       'Braking(m)', 'Grip(g)', 'Class', 'Type', 'Series', 'Purchase_Cost',
       'Service_Time(min)', 'Service_Cost(R$)', 'Availability'],
      dtype='object')

Here's the shape of the data frame: 

In [59]:
rr3_df.shape

(300, 14)

How about the general statistics of the features which hold numerical values (i.e. data type of `float64`)?

In [89]:
rr3_df.describe()

PR  Top_Speed(kph)  Acceleration(s)  Braking(m)     Grip(g)  \
count  300.000000      299.000000       299.000000  299.000000  299.000000   
mean    49.685000      310.745819         3.746656   30.746154    1.329130   
std     26.839124       46.426701         1.959458    5.603481    0.502347   
min      0.100000      180.000000         1.900000   14.300000    0.720000   
25%     34.800000      280.500000         2.870000   28.600000    1.010000   
50%     48.300000      317.000000         3.500000   30.400000    1.160000   
75%     66.250000      333.000000         4.090000   32.300000    1.495000   
max    130.300000      447.000000        30.900000   48.700000    4.000000   

           Series  Service_Time(min)  Service_Cost(R$)  
count  300.000000         299.000000        298.000000  
mean     2.116667         164.662207       1432.013423  
std      1.113788          61.868796        522.325678  
min      1.000000           1.000000          0.000000  
25%      1.000000         142.500000       1185.500000  
50%      2.000000         180.000000       1510.000000  
75%      3.000000         180.000000       1910.500000  
max      7.000000         325.000000       2192.000000

Whoa, look at that! That's lots of info we obtained just from one command! Well, it already explains itself!

What makes me wondering is that why the count value for several features is not the exact number, except for PR and Series.

Therefore, I want to find out which row(s) have missing values, possibly NaN values.

In [105]:
rr3_df[rr3_df['Top_Speed(kph)'].isnull()]

Manufacturer               Model    PR  Top_Speed(kph)  Acceleration(s)  \
216  Mercedes-AMG  GT R F1 Safety Car  43.2             NaN              NaN   

     Braking(m)  Grip(g) Class Type  Series Purchase_Cost  Service_Time(min)  \
216         NaN      NaN     R   FR       1           NaN                NaN   

     Service_Cost(R$)  Availability  
216               NaN          True

A-ha! That's the vehicle which causes the 'count' value for numerical ones have 1 less!

This indicates that Formula 1 Safety Car's properties have not been entered in the data frame yet.

In addition, count of `Service_Cost(R$)` is actually 2 less. Which might be the other vehicle whose service cost is missing?

In [106]:
rr3_df[rr3_df['Service_Cost(R$)'].isnull()]

Manufacturer               Model    PR  Top_Speed(kph)  Acceleration(s)  \
157    Koenigsegg            Agera RS  82.1           447.0             2.56   
216  Mercedes-AMG  GT R F1 Safety Car  43.2             NaN              NaN   

     Braking(m)  Grip(g) Class Type  Series Purchase_Cost  Service_Time(min)  \
157        28.3      1.6     S   MR       1     1000 Gold              180.0   
216         NaN      NaN     R   FR       1           NaN                NaN   

     Service_Cost(R$)  Availability  
157               NaN          True  
216               NaN          True

Ah yes! I remembered that Koenigsegg Agera RS was added in version 8.0 and I did not include the corresponding value. Perhaps I should consider filling the value there in the next update of this data frame...

OK, I've got this knowledge, let's continue motivating myself...

With `describe()` method, I've seen many helpful decimal values. Can I do the same with `numpy` methods? Let's try them on PR column:

In [123]:
PR = rr3_df['PR'].values
print("Count: {}, Mean: {:.3}, Std. dev.: {:.5}".format(PR.size, PR.mean(), PR.std()))

Count: 300, Mean: 49.7, Std. dev.: 26.794


In [124]:
print("Min: {}, Max: {}".format(PR.min(), PR.max()))

Min: 0.1, Max: 130.3


In [132]:
print("25th, 50th, 75th  precent values: {}".format(np.percentile(PR, [25, 50, 75])))

25th, 50th, 75th  precent values: [34.8  48.3  66.25]


Yep, they are all doable! By obtaining this information above, I'd like to know which vehicles have the lowest and highest PR values.

In [137]:
lowest_PR = rr3_df[rr3_df['PR'] == 0.1]
lowest_PR.loc[:, ['Manufacturer', 'Model', 'Purchase_Cost']]

Manufacturer                       Model Purchase_Cost
187        Mazda                        RX-3       50 Gold
220       Nissan  Skyline 2000 GT-R (KPGC10)       50 Gold
245      Porsche            911 Targa (1974)      53000 R$

The lowest PR value is shared by 3 vehicles, not one! Well, why do Mazda RX-3 and Nissan Skyline 2000 GT-R cost Gold to buy then instead of R$?

In [153]:
highest_PR = rr3_df[rr3_df['PR'] == 130.3]
highest_PR.iloc[:,[0,1,3,4,5,6,10]]

Manufacturer  Model  Top_Speed(kph)  Acceleration(s)  Braking(m)  Grip(g)  \
204      McLaren  MP4-X           402.0              1.9        14.3      4.0   

    Purchase_Cost  
204     1200 Gold

Wow! McLaren MP4-X must be immensely fast! This futuristic vehicle really deserves to have the highest PR in the game. Moreover, it has the highest acceleration, cornering grip and lowest braking distance!

On the other hand, w.r.t. top speed, it doesn't have the highest! So, which one?

In [161]:
highest_TS = rr3_df['Top_Speed(kph)'].max()
highest_TS_name = rr3_df[rr3_df['Top_Speed(kph)'] == highest_TS]
highest_TS_name[['Manufacturer', 'Model', 'Top_Speed(kph)']]

Manufacturer     Model  Top_Speed(kph)
157   Koenigsegg  Agera RS           447.0

Aah! The newly added Koenigsegg Agera RS has the highest top speed (447 kph) even with no upgrades!

Now, I wonder which vehicle appears the most throughout the career series.

In [158]:
highest_series = rr3_df['Series'].max()
rr3_df[rr3_df['Series'] == highest_series]

Manufacturer               Model    PR  Top_Speed(kph)  Acceleration(s)  \
263      Porsche  918 Spyder Concept  60.2           321.0              3.0   

     Braking(m)  Grip(g) Class Type  Series Purchase_Cost  Service_Time(min)  \
263        30.1     1.39     S   M4       7     845000 R$              215.0   

     Service_Cost(R$)  Availability  
263            1605.0          True

Ah yes, one of my favourite vehicle here: Porsche 918 Spyder Concept. Should this be a good choice to go for full upgrade?

OK, lastly, through these numbers above, I noticed that the minimum required servicing time is just 1 minute.
Thus, which vehicle does almost not need much servicing time?

In [162]:
minimum_service_time = rr3_df['Service_Time(min)'].min()
rr3_df[rr3_df['Service_Time(min)'] == minimum_service_time]

Manufacturer         Model   PR  Top_Speed(kph)  Acceleration(s)  \
217       Nissan  Silvia (S15)  1.7           243.0              5.5   

     Braking(m)  Grip(g) Class     Type  Series Purchase_Cost  \
217        34.4     0.85     P  FR(RHD)       3      25700 R$   

     Service_Time(min)  Service_Cost(R$)  Availability  
217                1.0             342.0          True

Hey, that's our good old friend Nissan Silvia! That was our first vehicle ever to own when we were newcomers in the game, right? Plus, it shouldn't be so expensive to get it as fully upgraded, though!

In contrast, now, which one has the highest servicing time?

In [163]:
max_service_time = rr3_df['Service_Time(min)'].max()
rr3_df[rr3_df['Service_Time(min)'] == max_service_time]

Manufacturer                Model    PR  Top_Speed(kph)  Acceleration(s)  \
299       Toyota  TS040 Hybrid (2014)  89.4           394.0              2.5   

     Braking(m)  Grip(g) Class Type  Series Purchase_Cost  Service_Time(min)  \
299        26.2     1.82     R   M4       2      950 Gold              325.0   

     Service_Cost(R$)  Availability  
299            2192.0          True

Ah, it's a Toyota! But, not a street legal vehicle; instead, from Endurance Motorsport Series. I figured out that it needs 5 hours 25 minutes to have it fully serviced. Likewise, the highest servicing cost belongs to this vehicle, too!

In [169]:
median_service_time = rr3_df['Service_Time(min)'].median()
print("Median of service time:", median_service_time)
rr3_df[rr3_df['Service_Time(min)'] == median_service_time].shape[0]

Median of service time: 180.0


138

OK, it's not surprising that the median value of servicing time is 180 minutes = 3 hours, since many new vehicles that were added through the latest updates, especially the expensive ones have their service times set to that value. Now, it's been raised to 138 vehicles!

For this reason, this time value is definitely dominant among all!

In [171]:
print("Ratio of vehicles having service time of 3 hours: {} percent".format(138/rr3_df.shape[0] * 100))

Ratio of vehicles having service time of 3 hours: 46.0 percent


By doing a simple calculation here, we imply that accounts for 46% of the vehicles.

One last thing which surprises me that there does exist one or more vehicles with NO service cost. Hmmm, what are those?

In [174]:
no_service_cost = rr3_df[rr3_df['Service_Cost(R$)'] == 0]
no_service_cost.loc[:,['Manufacturer', 'Model', 'Service_Time(min)']]

Manufacturer                                              Model  \
121    Formula 1                                     F1 Academy Car   
122    Formula 1                            Renault F1 Team R.S. 19   
123    Formula 1                               Red Bull Racing RB15   
124    Formula 1                                   Toro Rosso STR14   
125    Formula 1                              Scuderia Ferrari SF90   
126    Formula 1  Mercedes-AMG Petronas Motorsport F1 W10 EQ Power+   
127    Formula 1                                 Haas F1 Team VF-19   
128    Formula 1                              Alfa Romeo Racing C38   
129    Formula 1                              McLaren F1 Team MCL34   
130    Formula 1                          SPScore Racing Point RP19   
131    Formula 1                         Rokit Williams Racing FW42   

     Service_Time(min)  
121              180.0  
122              180.0  
123              180.0  
124              180.0  
125              180.0  
126              180.0  
127              180.0  
128              180.0  
129              180.0  
130              180.0  
131              180.0

Oh my goodness! These are the vehicles from 2019 season of Formula 1 which do not cost anything to service! I suppose there seem to be privileges on these vehicles since F1 series was the last one added through the motorsport series in the game! Anyway, they all share in common 3 hours of servicing time...

### **All right folks, that's all for the 1st outcome page for Vehicles in Real Racing 3!**

**In the next part, I will reveal more statistical info for better insight of this dataset; and putting some visualization charts will be nice here!**